tests/e2e/files/rulebooks/operators/test_selectattr_operator.yml

```diff
@@ -213,3 +222,9 @@
       action:
         debug:
           msg: "Output for testcase #11"
+
+    - name: selectattr and squared accessor interaction
+      condition: event.asd["x"][1][2] is selectattr("a.b", "==", 3.1415)
```

Comment:
Can we use the bracket accessor inside the selectattr? this is the thing that I am most concerned.

Reply:
I'm a bit confused: should it?

- in the ansible-rulebook doc, it only [mentions "a valid key"](https://ansible.readthedocs.io/projects/rulebook/en/stable/conditions.html#:~:text=The%20key%20is%20a%20valid%20key%20name%20in%20the%20object), so I would say actually I would _NOT_ expect that
- in the jinjia doc, there is never example of an expression (searching for "selectattr" in [this page](https://jinja.palletsprojects.com/en/3.1.x/templates/#jinja-filters.selectattr)) only of a "valid key", so I would say actually I would _NOT_ expect that

So I am putting forward this test showing the dot accessor still worked per previous behaviour (ie _before_ this PR), even if I can't find a reference that actually claims it is expected to be working.

👉  Is there a semantic definition of `selectattr` beyond the linked docs above I can reference to, please?

I'm happy to investigate if it's possible to make the syntax work as part of the string literal of the first argument of the selectattr, but I will keep the format of
```
id.something["else"]
```
etc.
In other words, like per ansible rulebook, you cannot start with a `[`.

Reply:
Sorry, I'm not sure If I'm misunderstanding something. From my understanding, if we support the bracket accessor users would try to do something like:
```
event.alerts is selectattr(labels["api.k8s.v1/namespace"], "==", "production")
```
Or probably 
```
event.alerts is selectattr('labels["api.k8s.v1/namespace"]', "==", "production")
```

I expected to support that kind of usecases. Am I wrong?

Reply:
I do see your point, and thank you indeed for raising it in the first place!

My understanding is however as I highlighted in https://github.com/ansible/ansible-rulebook/pull/539#discussion_r1239772983

In other words, just by looking at the documentation of ansible-rulebook and jinjia (see links I provided above), I would _not_ have expected even the simple dot-accessor to work as some existing tests are actually checking, [example here](https://github.com/ansible/ansible-rulebook/blob/0f5e4d4291691ec23c4c67d2751da7c9293b9716/tests/e2e/files/rulebooks/operators/test_selectattr_operator.yml#L119).

So, in this PR, I just kept existing behaviour nevertheless.

I'm open to discuss extending the `id.something["else"][0][-1]['x]` to other use-cases, but I would prefer if we ground the discussion based on some documentation/reference for the semantic to respect?

Reply:
As you wish. If it's a matter of a proper documentation, well, I admit that `The key is a valid key name in the object.` is not clear enough. We made that reading from the beginning and I'm just applying the same criteria here. 

If you consider that it is out of the scope, up to you. 

In that case can I ask for an update the of these docs for a better explanation to set properly the user expectation? Anyway this PR should include documentation to indicate the support for the new accessor. 

Also, it should include it in the changelog. https://github.com/ansible/ansible-rulebook/blob/main/CHANGELOG.md

Reply:
Sure; I propose this way: with regards to the impact/relation on `selectattr` specifically, I will investigate how that works ("empirically") in _play_book; since we have this doubt, it's worthy to investigate it at once.

Regardless, will anyway update doc and changelog note.

What do you think?

Reply:
> Sure; I propose this way: with regards to the impact/relation on `selectattr` specifically, I will investigate how that works ("empirically") in _play_book; since we have this doubt, it's worthy to investigate it at once.

I have taken some time to investigate as agreed ☝️ 

## official manual and source code

I believe `selectatrr` is "a bit underspecified" in the Ansible playbook [manual](https://docs.ansible.com/ansible/latest/playbook_guide/complex_data_manipulation.html#:~:text=of%20the%20condition.-,selectattr,-/rejectattr%3A%20very%20similar); that is, the behaviour is not described beyond referencing it to Jinja.

If I go to Jinja [manual](https://jinja.palletsprojects.com/en/3.1.x/templates/#jinja-filters.selectattr:~:text=applying%20a%20test%20to%20the-,specified%20attribute,-of%20each%20object%2C%20and%20only), it is "a bit underspecified" there in the ultimate sense nowhere in the docs I seem to find how the 1st string argument is supposed to behave (just a key? or a "dot accessor expression"?).

If I look at jinja source code I need to perform at least 3 hops before finding any mention if the 1st argument is directly a key or a "dot accessor expression" or else (from [here](https://github.com/pallets/jinja/blob/7b48764688dce68e99341bec59b1eedf0c5c3ba5/src/jinja2/filters.py#L1589) to [here](https://github.com/pallets/jinja/blob/7b48764688dce68e99341bec59b1eedf0c5c3ba5/src/jinja2/filters.py#L1757) to [here](https://github.com/pallets/jinja/blob/7b48764688dce68e99341bec59b1eedf0c5c3ba5/src/jinja2/filters.py#L1723) to [here](https://github.com/pallets/jinja/blob/7b48764688dce68e99341bec59b1eedf0c5c3ba5/src/jinja2/filters.py#L62-L63) )    
and appears to be actually manipulated by splitting on the `.`.

## empirically testing _play_book

I then checked empirically using Ansible _play_book, here: https://github.com/tarilabs/demo20230626-ansible-playground#readme

So on the playbook front I would conclude, while a bit underspecified maybe, I believe most people expect it to be using as direcly as a key attribute (in "full") as that anyway is what is reported by the official manual, and actually most 3rd party tutorial I could google online as I was learning more about this feature of Ansible myself 😅 

In other words, most manual/tutorial/examples I could find are in the form of:

```
selectattr("color", "==", "red") 
```

and I could _**not**_ find any manual/tutorial/examples in the form of:

```
selectattr("navigating.deep.inside", "==", "red") 
```

ie: no squared accessor in the 1st parameter of selectattr, but splitting on `.` "brutally"

## current behaviour of drools-ansible

In drools-ansible-rulebook-integration, the `extractMapAttribute` implements currently per jinja2/actual behaviour:
https://github.com/kiegroup/drools-ansible-rulebook-integration/blob/711ad015893c0cc1f469a793df886b5415e8d7e9/drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/domain/constraints/SelectAttrConstraint.java#L50

ie: no squared accessor in the 1st parameter of `selectattr`, but splitting on `.` "brutally" like playbook/jinja does.

---

So I believe on a behaviour point-of-view for rulebook as discussed here, we have two options:

## conclusions

1. **Option1**. leave current rulebook behaviour as-is, which mimics ansible playbook and jinja2 behaviour. In other words in `selectattr` use key (or at best a dot-expression). In other words, leave "underspecified" also in rulebook.

2. **Option2**. change rulebook behaviour for `selectattr` first argument to _extend_ to support squared accessors, etc. In other words diverge from ansible playbook and jinjia2 selectattr function behaviour.

If it was for me, I'd leave it at option1 without any further changes. If anything, I could make a documentation link in rulebook when speaking of selectattr, to point to Ansible _playbook_ documentation for the same function.

👉  @Alex-Izquierdo do you agree? I remain available with you and team if a quick call is preferred.

Reply:
Well, on one hand I really believe users will expect or try to use the new accessor there if it is supported, that was my original concern. 
For sure option one is the most reasonable and we are following the "emulation" of ansible for all these kind of things, so I'm not sure if we will want to implement it even if it could improve the functionality and UX. 

IMO, we can go ahead with option one and we can discuss it in later sync calls. 

Thank you again for this exhaustive and excellent contribution @tarilabs

Reply:
glad to contribute 🚀 👍